hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -441,6 +441,121 @@ struct FunctionVisitor : visitor::PreOrder<bool, FunctionVisitor> {
     }
 };
 
+struct TypeVisitor : visitor::PreOrder<bool, TypeVisitor> {
+    Stage _stage = Stage::COLLECT;
+    std::map<ID, bool> _used;
+
+    void collect(Node& node) {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    bool prune_uses(Node& node) { return false; }
+    bool prune_decls(Node& node) {
+        _stage = Stage::PRUNE_DECLS;
+
+        bool any_modification = false;
+
+        for ( auto i : this->walk(&node) )
+            if ( auto x = dispatch(i) )
+                any_modification = any_modification || *x;
+
+        return any_modification;
+    }
+
+    result_t operator()(const declaration::Type& x, position_t p) {
+        const auto type_id = x.typeID();
+
+        if ( ! type_id )
+            return false;
+
+        switch ( _stage ) {
+            case Stage::COLLECT:
+                // Record the type if not already known. If the type is part of an external API record it as used.
+                _used.insert({*type_id, x.linkage() == declaration::Linkage::Public});
```

Comment:
Yeah, this is different since here we only want to add an entry (record a use derived from the type visibility) if non previously exists. The other visitors always record a use. With that actual _uses of a type_ override whatever we derived from just _declarations of the type_.